File: android/app/src/main/java/net/bradball/teetimecaddie/android/ui/common/avatars/TtcAvatar.kt
package net.bradball.teetimecaddie.android.ui.common.avatars

import android.content.res.Configuration
import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxScope
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.drawBehind
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.graphics.painter.ColorPainter
import androidx.compose.ui.graphics.painter.Painter
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import net.bradball.teetimecaddie.android.theme.TtcColorRole
import net.bradball.teetimecaddie.android.theme.MyApplicationTheme

/**
 * A circular avatar from the Fairway Morning "Avatars & players" design system.
 *
 * `TtcAvatar` has three flavors, exposed as overloads over one shared circle:
 * - **initials** — a [color]-tinted circle with the player's [initials];
 * - **photo** — a caller-supplied, already-loaded image clipped to the circle (no color role);
 * - **placeholder** — the dashed [TtcAvatarPlaceholder.Guest] / [TtcAvatarPlaceholder.Empty] affordance.
 *
 * [size] is a free [Dp] (see [TtcAvatarDefaults] for the named sizes) so the same view serves the
 * stack (28dp) and the photo picker (88dp) as well as standalone use.
 *
 * @param initials The text to show (usually one or two letters).
 * @param modifier Modifier for the avatar.
 * @param color The [TtcColorRole] the initials disc is tinted with. Defaults to [TtcColorRole.Primary].
 * @param size The avatar diameter. Defaults to [TtcAvatarDefaults.MediumSize].
 */
@Composable
fun TtcAvatar(
    initials: String,
    modifier: Modifier = Modifier,
    color: TtcColorRole = TtcColorRole.Primary,
    size: Dp = TtcAvatarDefaults.MediumSize,
) {
    val (background, content) = TtcAvatarDefaults.toneColors(color)
    AvatarCircle(modifier = modifier, size = size, background = background) {
        Text(
            text = initials,
            color = content,
            style = TtcAvatarDefaults.textStyle(size),
            maxLines = 1,
        )
    }
}

/**
 * A [TtcAvatar] that shows a supplied [photo] (already-loaded [Painter]) clipped to the circle. There
 * is no tone — the image fills the avatar.
 *
 * @param photo The image to display, cropped to fill the circle.
 * @param modifier Modifier for the avatar.
 * @param size The avatar diameter. Defaults to [TtcAvatarDefaults.MediumSize].
 */
@Composable
fun TtcAvatar(
    photo: Painter,
    modifier: Modifier = Modifier,
    size: Dp = TtcAvatarDefaults.MediumSize,
) {
    AvatarCircle(modifier = modifier, size = size, background = Color.Transparent) {
        Image(
            painter = photo,
            contentDescription = null,
            contentScale = ContentScale.Crop,
            modifier = Modifier.matchParentSize(),
        )
    }
}

/**
 * A [TtcAvatar] placeholder affordance — a dashed circle with a person ([TtcAvatarPlaceholder.Guest])
 * or add ([TtcAvatarPlaceholder.Empty]) glyph. Ignores tone.
 *
 * @param placeholder Which placeholder to render.
 * @param modifier Modifier for the avatar.
 * @param size The avatar diameter. Defaults to [TtcAvatarDefaults.MediumSize].
 */
@Composable
fun TtcAvatar(
    placeholder: TtcAvatarPlaceholder,
    modifier: Modifier = Modifier,
    size: Dp = TtcAvatarDefaults.MediumSize,
) {
    val scheme = MaterialTheme.colorScheme
    val border = scheme.outlineVariant
    val dash = TtcAvatarDefaults.dashPathEffect()
    Box(
        modifier = modifier
            .size(size)
            .drawBehind {
                val stroke = TtcAvatarDefaults.BorderWidth.toPx()
                drawCircle(
                    color = border,
                    radius = (this.size.minDimension - stroke) / 2f,
                    style = Stroke(width = stroke, pathEffect = dash),
                )
            },
        contentAlignment = Alignment.Center,
    ) {
        Icon(
            painter = TtcAvatarDefaults.placeholderIcon(placeholder).painter,
            contentDescription = null,
            tint = scheme.onSurfaceVariant,
            modifier = Modifier.size(TtcAvatarDefaults.iconSize(size)),
        )
    }
}

@Composable
private fun AvatarCircle(
    modifier: Modifier,
    size: Dp,
    background: Color,
    content: @Composable BoxScope.() -> Unit,
) {
    Box(
        modifier = modifier
            .size(size)
            .clip(CircleShape)
            .background(background),
        contentAlignment = Alignment.Center,
        content = content,
    )
}

@Preview(name = "Light", showBackground = true)
@Preview(name = "Dark", showBackground = true, uiMode = Configuration.UI_MODE_NIGHT_YES)
@Composable
private fun TtcAvatarPreview() {
    MyApplicationTheme {
        Surface {
            Column(
                modifier = Modifier.size(width = 320.dp, height = 260.dp),
                verticalArrangement = Arrangement.spacedBy(16.dp),
            ) {
                Row(
                    horizontalArrangement = Arrangement.spacedBy(12.dp),
                    verticalAlignment = Alignment.CenterVertically,
                    modifier = Modifier.padding(16.dp),
                ) {
                    TtcAvatar(initials = "B", color = TtcColorRole.Secondary)
                    TtcAvatar(initials = "M", color = TtcColorRole.Primary)
                    TtcAvatar(initials = "J", color = TtcColorRole.Tertiary)
                    TtcAvatar(initials = "C", color = TtcColorRole.Error)
                    TtcAvatar(initials = "K", color = TtcColorRole.Neutral)
                    TtcAvatar(photo = ColorPainter(MaterialTheme.colorScheme.tertiaryContainer))
                    TtcAvatar(placeholder = TtcAvatarPlaceholder.Guest)
                    TtcAvatar(placeholder = TtcAvatarPlaceholder.Empty)
                }
                Row(
                    horizontalArrangement = Arrangement.spacedBy(12.dp),
                    verticalAlignment = Alignment.CenterVertically,
                    modifier = Modifier.padding(horizontal = 16.dp),
                ) {
                    TtcAvatar(initials = "D", color = TtcColorRole.Primary, size = TtcAvatarDefaults.SmallSize)
                    TtcAvatar(initials = "D", color = TtcColorRole.Primary, size = TtcAvatarDefaults.MediumSize)
                    TtcAvatar(initials = "D", color = TtcColorRole.Primary, size = TtcAvatarDefaults.LargeSize)
                }
            }
        }
    }
}

File: ios/TeeTimeCaddie/TeeTimeCaddie/Source/ui/common/Avatars/TtcAvatar.swift
import SwiftUI
import ThemeUI

/// A circular avatar from the Fairway Morning "Avatars & players" design system.
///
/// `TtcAvatar` has three flavors, exposed as separate initializers over one shared circle:
/// - **initials** — a `color`-tinted circle with the player's initials;
/// - **photo** — a caller-supplied, already-loaded `Image` clipped to the circle (no color role);
/// - **placeholder** — the dashed `.guest` / `.empty` affordance.
///
/// `size` is a free `CGFloat` (see ``TtcAvatarStyle`` for the named sizes) so the same view serves the
/// stack (28), the photo picker (88), and standalone use.
struct TtcAvatar: View {
    @EnvironmentObject private var theme: AppTheme

    private enum Content {
        case initials(String, TtcColorRole)
        case photo(Image)
        case placeholder(TtcAvatarPlaceholder)
    }

    private let content: Content
    private let size: CGFloat

    /// An initials avatar tinted by `color`.
    init(_ initials: String, color: TtcColorRole = .primary, size: CGFloat = TtcAvatarStyle.mediumSize) {
        self.content = .initials(initials, color)
        self.size = size
    }

    /// A photo avatar — the supplied image cropped to fill the circle. No color role.
    init(photo: Image, size: CGFloat = TtcAvatarStyle.mediumSize) {
        self.content = .photo(photo)
        self.size = size
    }

    /// A dashed placeholder avatar (`.guest` person / `.empty` add). Ignores the color role.
    init(placeholder: TtcAvatarPlaceholder, size: CGFloat = TtcAvatarStyle.mediumSize) {
        self.content = .placeholder(placeholder)
        self.size = size
    }

    var body: some View {
        avatarBody(theme.colorScheme)
            .frame(width: size, height: size)
    }

    @ViewBuilder
    private func avatarBody(_ scheme: ThemeColors) -> some View {
        switch content {
        case let .initials(initials, color):
            TtcAvatarInitials(initials: initials, colors: TtcAvatarStyle.toneColors(color, scheme), size: size)
        case let .photo(image):
            image
                .resizable()
                .scaledToFill()
                .frame(width: size, height: size)
                .clipShape(Circle())
        case let .placeholder(placeholder):
            TtcAvatarPlaceholderView(placeholder: placeholder, scheme: scheme, size: size)
        }
    }
}

fileprivate struct TtcAvatarInitials: View {
    let initials: String
    let colors: (background: Color, foreground: Color)
    let size: CGFloat

    var body: some View {
        Circle()
            .fill(colors.background)
            .overlay {
                Text(initials)
                    .font(.system(size: TtcAvatarStyle.textSize(size), weight: .semibold))
                    .foregroundStyle(colors.foreground)
            }
    }
}

fileprivate struct TtcAvatarPlaceholderView: View {
    let placeholder: TtcAvatarPlaceholder
    let scheme: ThemeColors
    let size: CGFloat

    var body: some View {
        Circle()
            .strokeBorder(scheme.outlineVariant, style: TtcAvatarStyle.dashedStroke())
            .overlay {
                Image(systemName: TtcAvatarStyle.placeholderSymbol(placeholder))
                    .font(.system(size: TtcAvatarStyle.iconSize(size)))
                    .foregroundStyle(scheme.onSurfaceVariant)
            }
    }
}

#Preview("Light") {
    TeeTimeCaddieTheme {
        AvatarPreviewContent()
    }
}

#Preview("Dark") {
    TeeTimeCaddieTheme {
        AvatarPreviewContent()
    }
    .preferredColorScheme(.dark)
}

private struct AvatarPreviewContent: View {
    var body: some View {
        VStack(alignment: .leading, spacing: 16) {
            HStack(spacing: 12) {
                TtcAvatar("B", color: .secondary)
                TtcAvatar("M", color: .primary)
                TtcAvatar("J", color: .tertiary)
                TtcAvatar("C", color: .error)
                TtcAvatar("K", color: .neutral)
                TtcAvatar(placeholder: .guest)
                TtcAvatar(placeholder: .empty)
            }
            HStack(spacing: 12) {
                TtcAvatar("D", color: .primary, size: TtcAvatarStyle.smallSize)
                TtcAvatar("D", color: .primary, size: TtcAvatarStyle.mediumSize)
                TtcAvatar("D", color: .primary, size: TtcAvatarStyle.largeSize)
            }
        }
        .padding()
    }
}
